tools/net-common.in

```diff
@@ -63,6 +63,68 @@ function fix_mac {
   ip link set $INTERFACE address $FIXED_MAC || true
 }
 
+function bridge_vlan_add {
+  local VID="${1}"
+  shift
+
+  # when the VID is in use by traditional VLAN interface, then the
+  # tap on the VLAN aware bridge will not receive traffic
+  if [ -r /proc/net/vlan/config ]; then
+    if ! awk -F '[| ]*' -v vlan="^${VID}$" 'match($2, vlan) { exit 1 }' /proc/net/vlan/config; then
+      echo "VLAN ${VID} is in use by interface $(awk -F '[| ]*' -v vlan="^${VID}$" 'match($2, vlan) { print $1 }' /proc/net/vlan/config)"
+      exit 1
+    fi
+  fi
+
+  # add the VLAN to the tap
+  # $@ is either empty or contains "pvid untagged"
+  bridge vlan add dev ${INTERFACE} vid ${VID} "${@}" master
+  # add the VLAN to the bridge uplinks
+  local uplink
+  for uplink in ${BRIDGE_UPLINKS}; do
+    bridge vlan add dev ${uplink} vid ${VID} master
+  done
+}
+
+function setup_bridge_vlan_aware {
+  # enforce the admin to configure vlan filtering explicit
+  if [ $(</sys/class/net/${LINK}/bridge/vlan_filtering) -ne 1 ]; then
+    echo "vlan filtering must be enabled for the bridge ${LINK}"
```

Comment:
I think this message should also say *why* vlan_filtering must be enabled, i.e. something along the lines "Instance blah NIC foo configured for VLAN support, but bridge ${LINK} has VLAN filtering disabled".